vendor/bat-native-ads/src/bat/ads/ad_history_detail.cc

```diff
@@ -0,0 +1,100 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/ad_history_detail.h"
+
+#include "bat/ads/ad_content.h"
+#include "bat/ads/category_content.h"
+
+#include "bat/ads/internal/json_helper.h"
+
+#include "base/logging.h"
+
+namespace ads {
+
+AdHistoryDetail::AdHistoryDetail() :
+    timestamp_in_seconds(0),
+    uuid(""),
+    ad_content(),
+    category_content() {}
+
+AdHistoryDetail::AdHistoryDetail(const AdHistoryDetail& detail) :
+    timestamp_in_seconds(detail.timestamp_in_seconds),
+    uuid(detail.uuid),
+    ad_content(detail.ad_content),
+    category_content(detail.category_content) {}
+
+AdHistoryDetail::~AdHistoryDetail() = default;
+
+const std::string AdHistoryDetail::ToJson() const {
+  std::string json;
+  SaveToJson(*this, &json);
+  return json;
+}
+
+Result AdHistoryDetail::FromJson(
+    const std::string& json,
+    std::string* error_description) {
+  rapidjson::Document document;
+  document.Parse(json.c_str());
+
+  if (document.HasParseError()) {
+    if (error_description != nullptr) {
+      *error_description = helper::JSON::GetLastError(&document);
+    }
+
+    return FAILED;
+  }
+
+  if (document.HasMember("timestamp_in_seconds")) {
+    timestamp_in_seconds = document["timestamp_in_seconds"].GetUint64();
+  }
+
+  if (document.HasMember("uuid")) {
+    uuid = document["uuid"].GetString();
+  }
+
+  if (document.HasMember("ad_content")) {
+    rapidjson::StringBuffer buffer;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+    const auto& value = document["ad_content"];
+    if (!value.Accept(writer) ||
```

Comment:
@tmancey I think there must be a better way to read this value than having to use the extra StringBuffer. Is this how it's meant to work or is there something simpler I can do?

Reply:
We can beautify the code with base::jsonreader/writer, so good for now